openedx/core/djangoapps/course_groups/management/commands/tests/test_post_cohort_membership_fix.py

```diff
@@ -0,0 +1,74 @@
+"""
+Test for the post-migration fix commands that are included with this djangoapp
+"""
+from django.core.management import call_command
+from django.test.client import RequestFactory
+
+from openedx.core.djangoapps.course_groups.views import cohort_handler
+from openedx.core.djangoapps.course_groups.cohorts import get_cohort_by_name
+from openedx.core.djangoapps.course_groups.tests.helpers import config_course_cohorts
+from openedx.core.djangoapps.course_groups.models import CohortMembership
+from student.tests.factories import UserFactory
+from xmodule.modulestore.tests.django_utils import ModuleStoreTestCase
+from xmodule.modulestore.tests.factories import CourseFactory
+
+
+class TestPostMigrationFix(ModuleStoreTestCase):
+    """
+    Base class for testing post-migration fix commands
+    """
+    def setUp(self):
+        """
+        setup course, user and request for tests
+        """
+        super(TestPostMigrationFix, self).setUp()
+        self.course1 = CourseFactory.create()
+        self.course2 = CourseFactory.create()
+        self.user1 = UserFactory(is_staff=True)
+        self.user2 = UserFactory(is_staff=True)
+        self.request = RequestFactory().get("dummy_url")
+        self.request.user = self.user1
+
+    def test_post_cohortmembership_fix(self):
+        """
+        Test that changes made *after* migration, but *before* turning on new code are handled properly
+        """
+        # First, we're going to simulate some problem states that can arise during this window
+        config_course_cohorts(self.course1, is_cohorted=True, auto_cohorts=["Course1AutoGroup1", "Course1AutoGroup2"])
+
+        # Get the cohorts from the courses, which will cause auto cohorts to be created
+        cohort_handler(self.request, unicode(self.course1.id))
+        course_1_auto_cohort_1 = get_cohort_by_name(self.course1.id, "Course1AutoGroup1")
+        course_1_auto_cohort_2 = get_cohort_by_name(self.course1.id, "Course1AutoGroup2")
+
+        # When migrations were first run, the users were assigned to CohortMemberships correctly
+        membership1 = CohortMembership(
+            course_id=course_1_auto_cohort_1.course_id,
+            user=self.user1,
+            course_user_group=course_1_auto_cohort_1
+        )
+        membership1.save()
+        membership2 = CohortMembership(
+            course_id=course_1_auto_cohort_1.course_id,
+            user=self.user2,
+            course_user_group=course_1_auto_cohort_1
+        )
+        membership2.save()
+
+        # But before CohortMembership code was turned on, some changes were made:
+        course_1_auto_cohort_2.users.add(self.user1)  # user1 is now in 2 cohorts in the same course!
+        course_1_auto_cohort_2.users.add(self.user2)
+        course_1_auto_cohort_1.users.remove(self.user2)  # and user2 was moved, but no one told CohortMembership!
+
+        # run the post-CohortMembership command
+        call_command('post_cohort_membership_fix', 'commit')
```

Comment:
Maybe before this, you could test the dry run version and ensure nothing gets changed?